src/tokens/user_token.rs

```
@@ -67,12 +75,28 @@ impl TwitchToken for UserToken {
 
     fn login(&self) -> Option<&str> { self.login.as_deref() }
 
-    async fn refresh_token<RE, C, F>(&mut self, _: C) -> Result<(), RefreshTokenError<RE>>
+    async fn refresh_token<RE, C, F>(
+        &mut self,
+        http_client: C,
+    ) -> Result<(), RefreshTokenError<RE>>
     where
         RE: std::error::Error + Send + Sync + 'static,
         C: FnOnce(HttpRequest) -> F,
-        F: Future<Output = Result<HttpResponse, RE>>, {
-        Err(RefreshTokenError::NoRefreshToken)
+        F: Future<Output = Result<HttpResponse, RE>>,
+    {
+        if let Some(client_secret) = self.client_secret.clone() {
+            let (access_token, expires, refresh_token) = if let Some(token) =
+                self.refresh_token.take()
+            {
+                crate::refresh_token(http_client, token, &self.client_id, &client_secret).await?
+            } else {
+                return Err(RefreshTokenError::NoRefreshToken);
+            };
+            self.access_token = access_token;
+            self.expires = expires;
+            self.refresh_token = refresh_token;
+        }
+        Ok(())
```

If no secret is available this should probably return a new error `RefreshTokenError::NoClientSecret`